pkg/core/imageprocessing/resize.go

```diff
@@ -0,0 +1,281 @@
+package imageprocessing
+
+import (
+	"github.com/davidbyttow/govips/pkg/vips"
+)
+
+type ResizeScalingMode string
+
+func (m ResizeScalingMode) ShouldInferAnotherSideWithAspectRatio() bool {
+	switch m {
+	case ResizeScalingModeLfit:
+		return true
+	case ResizeScalingModeMfit:
+		return true
+	case ResizeScalingModeFill:
+		return false
+	case ResizeScalingModePad:
+		return false
+	case ResizeScalingModeFixed:
+		return true
+	default:
+		panic("unreachable")
+	}
+}
+
+const (
+	ResizeScalingModeLfit  ResizeScalingMode = "lfit"
+	ResizeScalingModeMfit  ResizeScalingMode = "mfit"
+	ResizeScalingModeFill  ResizeScalingMode = "fill"
+	ResizeScalingModePad   ResizeScalingMode = "pad"
+	ResizeScalingModeFixed ResizeScalingMode = "fixed"
+)
+
+const ResizeScalingModeDefault = ResizeScalingModeLfit
+
+var ResizeDefaultColor = ColorWhite
+
+type Resize struct {
+	ScalingMode ResizeScalingMode
+	Width       int
+	Height      int
+	LongerSide  int
+	ShorterSide int
+	Color       Color
+}
+
+var _ Operation = &Resize{}
+
+// nolint: gocyclo
+func (o *Resize) Apply(ctx *OperationContext) error {
+	originalWidth := ctx.Image.Width()
+	originalHeight := ctx.Image.Height()
+	aspectRatio := ratio(originalWidth, originalHeight)
+
+	targetWidth := 0
+	targetHeight := 0
+	var longerTargetSide *int
+	var shorterTargetSide *int
+
+	if originalWidth > originalHeight {
+		longerTargetSide = &targetWidth
+		shorterTargetSide = &targetHeight
+	} else {
+		longerTargetSide = &targetHeight
+		shorterTargetSide = &targetWidth
+	}
+
+	// Set targetWidth and targetHeight if they are specified directly.
+	if o.LongerSide != 0 {
+		*longerTargetSide = o.LongerSide
+	}
+	if o.ShorterSide != 0 {
+		*shorterTargetSide = o.ShorterSide
+	}
+	if o.Width != 0 {
+		targetWidth = o.Width
+	}
+	if o.Height != 0 {
+		targetHeight = o.Height
+	}
+
+	// w h l s are not specified, w and h resolve to original size.
+	if targetWidth == 0 && targetHeight == 0 {
+		targetWidth = originalWidth
+		targetHeight = originalHeight
+	}
+
+	// w is not specified
+	if targetWidth == 0 {
+		if o.ScalingMode.ShouldInferAnotherSideWithAspectRatio() {
+			targetWidth = roundFloat(float64(targetHeight) * aspectRatio)
+		} else {
+			targetWidth = targetHeight
```

Comment:
It is by definition.

For some scaling mode, it makes more sense to have `w` default to `h`.

For example, for resizing mode `fill`, if only width is specified, it makes more sense to have height equal to width. If height is equal to the width divided by the aspect ratio, you cannot see the `fill` effect because the aspect ratio is the same.

Reply:
Okay. I think we need to note this down in the spec, that the default value of `w` and `h` depends on the scaling mode and aspect ratio.

Reply:
https://github.com/SkygearIO/features/pull/360/commits/bb95bfffa2f0c13c91906acc869bb0f52619e517